src/main/java/teammates/logic/core/FeedbackSessionsLogic.java

```diff
@@ -492,22 +492,18 @@ public void restoreFeedbackSessionFromRecycleBin(String feedbackSessionName, Str
      * Gets the expected number of submissions for a feedback session.
      */
     public int getExpectedTotalSubmission(FeedbackSessionAttributes fsa) {
-        List<StudentAttributes> students = studentsLogic.getStudentsForCourse(fsa.getCourseId());
-        List<InstructorAttributes> instructors = instructorsLogic.getInstructorsForCourse(fsa.getCourseId());
-        List<FeedbackQuestionAttributes> questions =
-                fqLogic.getFeedbackQuestionsForSession(fsa.getFeedbackSessionName(), fsa.getCourseId());
-        List<FeedbackQuestionAttributes> studentQns = fqLogic.getFeedbackQuestionsForStudents(questions);
+        int numOfStudents = studentsLogic.getNumberOfStudentsForCourse(fsa.getCourseId());
+        List<String> instructorEmails = instructorsLogic.getInstructorEmailsForCourse(fsa.getCourseId());
+        Map<FeedbackParticipantType, Integer> giverTypeCounts = fqLogic.getFeedbackQuestionGiverCountForSession(
+                fsa.getFeedbackSessionName(), fsa.getCourseId());
 
         int expectedTotal = 0;
-
-        if (!studentQns.isEmpty()) {
-            expectedTotal += students.size();
+        if (fqLogic.getFeedbackQuestionForStudentCountFrom(giverTypeCounts) > 0) {
```

Comment:
I just realized, you should be able to make use of `hasFeedbackQuestionsForStudents` here

Reply:
You are right... I just made those methods

Reply:
I just removed the `giverTypeCount` method and replace with overloaded `hasQnForX`

Reply:
I don't think even the overloaded method is needed

Reply:
The other `hasFeedbackQuestionsForInstructors` has `isCreatorOfSession` which in turn fetches the `FeedbackSession` entity per call. I don't feel like this is justified when we literally have the `fsa` available in the caller.

Reply:
Then, that other method is what you should modify

Reply:
Even then:
- Note that any attribute object can be easily created through means other than fetching from DB. As much as it may seem "wasted", fetching the item every time is far less fault-tolerant.
- In addition, you should notice that calling `hasFeedbackQuestionsForInstructors` for each instructor is in itself already wasted. Only one kind of instructor can possibly produce different outcome and that is the creator instructor.

Reply:
You are right about this. I should just eliminate down to only necessary information needed from db.

> Note that any attribute object can be easily created through means other than fetching from DB. As much as it may seem "wasted", fetching the item every time is far less fault-tolerant.

My take on this is that we want a snapshot of a session from one point in time. Fetching the session entity multiple times along the way might provide some inconsistency instead IMO.
Any ways we can guard against session entity not fetched from DB?

Reply:
Sorry I forgot to reply this.

Probably I shouldn't mention the first reason. It is definitely a concern, but another factor being in play is what the method is supposed to do. For instance, do you want a method that "does something with this given Attribute object", or "does something with this course ID and session name"? They can very well do different things even though in a glance they sound the same.
But determining which one is the right one to use is not a black-and-white matter. It is highly situational and depends on the business logic at the point of the method call. Which is why, "[session] entity not fetched from DB" is not even necessarily a problem.

But the second reason still stands as to why there is still room for even more optimization here.